### PR TITLE
Refactoring Usage Simulation + Abortion

### PR DESCRIPTION
**Pull request by klinakuf**
In this refactoring PR I intend to make several important changes/refactorings to the Usage Simulation extension: 

A prerequisite is the merge of stack fixes proposed by @stiesssh in #72. 

### 1. Dropping the parent/child relationship for the UserInterpretationContext. 

The change always puts the simulated user in one single interpretation context. There is no parent/child relationship as this is unnecessary. In the case of interpreting branches and loops, the UsageScenarioBehaviorContext is updated accordingly and there we have the parent/child relationship between a RootScenarioBehavior and, for example, a BranchScenarioBehavior. 

To continue the interpretation, the UserInterpretationProgressed event (which already existed but was not used) is thrown to Stop actions of inner scenarios of loops and branches. 

### 2. The stack is emptied when the user is aborted. 

If a user request is aborted, a reconciliation of the stack is necessary, i.e., in every corresponding extension, the frames have to be popped accordingly. Since this reconciliation is not present at the moment and may be somewhat challenging with choices on the semantic, in this PR, when a request fails, the complete user is aborted, and a new user replaces the old user with a new stack. The scenario is restarted, and the new user starts from the beginning and does not continue to interpret the leftover part.   

#### 3.1 onClosedWorkloadUserInitiated, 
the method was not doing what it promised in the comments and how it is supposed to work, this caused a bug while no stack frame was pushed for the user.

### CODE REVIEW

File: org.palladiosimulator.analyzer.slingshot.behavior.usagesimulation.data/src/org/palladiosimulator/analyzer/slingshot/behavior/usagemodel/entities/interpretationcontext/UserInterpretationContext.java
Comment (stiesssh) [inline org.palladiosimulator.analyzer.slingshot.behavior.usagesimulation.data/src/org/palladiosimulator/analyzer/slingshot/behavior/usagemodel/entities/interpretationcontext/UserInterpretationContext.java:1]: this is where you removed the parent, right? then add yourself to the authors :D

**Review by stiesssh — COMMENTED** [review]
- ran it on RMCU, got results. 
  * checked, that stack is empty after each normal request. 
  * did not check state of stack in case of abortion, as i don't know how to trigger abortion (deterministically)
 - changes appear to be reasonable.
 
 as already mentioned in person, i dislike creating new users in case of abortion with closed workload, but i also undestand, that it is difficult to solve in any other way right now, thus i shall not hinder the merging of this PR. But, i want a follow up issue, to takle that problem. maybe we even need to rethink how to identify probes.

**Review by stiesssh — CHANGES_REQUESTED** [review]
please merge `master` into this branch, i keep getting an assertion-error.

**Comment by klinakuf** [comment]
Is this the issue you wanted to have https://github.com/PalladioSimulator/Palladio-Analyzer-Slingshot-Extension-PCM-Core/issues/78

**Comment by stiesssh** [comment]
I tested with a more complicated UsageModel, that contains loops and branches: 

image
 
* I meassure *state of active resource*.
* Expectation: 4 times state = 1 $\rightarrow$ state = 0 for three seconds  $\rightarrow$ 4 times state = 1  $\rightarrow$ ...

* Reality: 
image

* Problem: Loops are broken. I tried different numbers of loop iterations, but the loop body always get executed only once.
* I'm not sure whether this Problem was caused by this PR, or whether it existed before. please check and either fix it (if related to this PR) or create a new issue (if not related to this PR).

**Review by stiesssh — COMMENTED** [review]
see comment.

**Review by stiesssh — CHANGES_REQUESTED** [review]
I keep getting problems with the preview feature you used. 
In the end i think it is cumbersome to adapt the IDE and the runconfig, especially if we get some student to work on it. 

Thus, i suggest to go back to default java 17 features.

**Review by stiesssh — CHANGES_REQUESTED** [review]
Now it's doing one iteration to much.

Or do i misunderstand the loop count? Cause i assume, that loop count = n means that, the loop body should be executed n-times, how ever now it get's executed n+1 times.

EDIT: to make sure, i executed the same model with SimuLizar, and the results are indeed different: 
**SimuLizar** (always 4 calls):
image


**Slingshot** (4 or 5 calls):
image

**Review by stiesssh — APPROVED** [review]
It works now.

Don't know what caused the problem before, probably some updated problems inside eclipse.

**Pull request merged**